core/src/main/scala/org/apache/spark/api/r/RRunner.scala

```diff
@@ -91,28 +91,70 @@ private[spark] class RRunner[U](
     }
 
     try {
-      return new Iterator[U] {
-        def next(): U = {
-          val obj = _nextObj
-          if (hasNext) {
-            _nextObj = read()
-          }
-          obj
+      newReaderIterator(dataStream, errThread)
+    } catch {
+      case e: Exception =>
+        throw new SparkException("R computation failed with\n " + errThread.getLines(), e)
+    }
+  }
+
+  protected def newReaderIterator(
```

Comment:
Codes here are almost not changed. I just exposed some method to inherit it in ArrowRRunner. This code needs refactoring as a common parent like PythonRunner but I don't target it here.